File: DSAlgorithms/Leetcode/ts/src/binary_search/33-search_rotated_arr.ts
function search(nums: number[], target: number): number {
    let l = 0;
    let r = nums.length -1;
    while(l <= r){
        let m = Math.floor((l + r) /2);
        if(target === nums[m])
            return m;
        if(nums[l] <= nums[m]){
            if(target > nums[m] || target < nums[l])     
                l = m + 1;
            else
                r = m - 1;
        }
        else {
            if(target < nums[m] || target > nums[r])
                r = m - 1;
            else l = m + 1;
        }
    }
    return -1;
};

File: dsa_algo/leetcode/binary_search/assorted/c++/33-search_rotated_arr.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
       int left = 0;
       int right = nums.size() - 1;
       while(left <= right) 
       {
            int mid = left + (right - left) /2 ;
            if(target == nums[mid]) //left sorted portion
                return mid;
            if(nums[left] <= nums[mid])
                if(target > nums[mid] || target < nums[left])
                    left = mid + 1;
                else 
                    right = mid - 1;
            else  //right sorted portion
                if(target < nums[mid] || target > nums[right])
                    right = mid - 1;
                else 
                    left = mid + 1;
       }
       return -1;
    }
};